tests/cypress/e2e/hedy_page/quiz.cy.js

```diff
@@ -0,0 +1,59 @@
+import { createJSDocTypeExpression } from "typescript";
+import { goToHedyPage } from "../tools/navigation/nav";
+
+beforeEach(() => {
+  goToHedyPage();
+  cy.get('div[data-tab="quiz"]').click();
+});
+
+/**
+ * This test is quite something. It randomly clicks through the quiz until it is
+ * finished to make sure the quiz can be completed. This is complicated by the
+ * fact that we can't know in advance whether we'll have click a correct or incorrect
+ * answer, and so we won't know whether what page we need to expect next.
+ *
+ * Cypress isn't actually designed to support test cases like this.
+ * Nevertheless, I don't want to hardcode questions and correct answers here, so
+ * we do a bit of complex work.
+ *
+ * Each of the pages renders a specific element with a page counter index,
+ * which advances by 1 every page. We can therefore wait for that element to
+ * appear, and then we'll know the page has completed async rendering. After
+ * we know the page has rendered, we can inspect what's visible to us, and
+ * do an appropriate thing (declare victory, click the perfunctory "next" button,
+ * or click a random answer).
+ */
+it('can complete the quiz by randomly clicking buttons', () => {
+  cy.get('*[data-cy="start-quiz"]').click();
+
+  randomlyClickThroughUntilFinished(0);
+
+  // We must use a recursive function to loop, because of the async nature of cypress
+  function randomlyClickThroughUntilFinished(expectedCounter) {
+    // This will wait until the page with the counter value has loaded
+    cy.get(`input[data-cy="ctr:${expectedCounter}"]`);
+
+    // Now that the page has loaded, it's safe to inspect it and do a conditional thing
+    // depending on what it looks like.
+    cy.get('*[data-tabtarget="quiz"]').then(($quizpane) => {
+      if ($quizpane.find('*[data-cy="quiz-done"]').length) {
+        // Yay, finished!
+        return;
+      }
+
+      if ($quizpane.find('*[data-cy="advance-quiz"]').length) {
+        // Click "next question" button"
+        cy.get('*[data-cy="advance-quiz"]').click();
+        randomlyClickThroughUntilFinished(expectedCounter + 1);
+        return;
+      }
+
+      // Otherwise click a random answer
+      cy.get('*[data-cy="quiz-choice"]:not(.incorrect-option)').then(($buttons) => {
+        return Cypress._.sample($buttons.toArray());
+      }).click();
+      cy.get('button.pick-answer-button:visible').click();
+      randomlyClickThroughUntilFinished(expectedCounter + 1);
```

Comment:
Love this, so the recursivity prevents the use of cy.wait()? Creative :D

Reply:
The `cy.wait()` is normally there because we need to wait until the page refresh is definitely "done", and we guess at a time in which that swap has happened. If we know that the current page **definitely does not** contain a particular element, and the page we're waiting for **definitely does have** a particular element, we can use `cy.get(<that element>)`, and it will finish as soon as the element is found (or wait up until a maximum of 4 seconds before giving up).

In our case, the page counter prevents the use of `cy.wait()`.  On every page load, there will be hidden element with `data-cy="ctr:0"`, `data-cy="ctr:1"`, `data-cy="ctr:2"`, etc. If we are currently on page `ctr:1`, then we know that we can wait for `ctr:2` to become available to know that the new content has been swapped in.

--------

The recursion is because we can't use a `while()` loop because of the async nature of Cypress. All cypress commands aren't executed immediately, they are queued and execute "later", when the page has finished refreshing. That means while the `while()` loop is executing, nothing has loaded yet. We must do our actual work inside the `.then(() => { ... })` callback, and the only way to loop in those kinds of situations is to use recursion to start a new async operation.